### PR TITLE
Reference roi align 3 implementation

### PR DESCRIPTION
**Pull request by blesniewski**

### CODE REVIEW

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 1]
```diff
@@ -0,0 +1,243 @@
...
+                                                                      static_cast<T>(0.5f));
+
+                                    if (sample_x < -1.0 || sample_x > feature_map_width ||
+                                        sample_y < -1.0 || sample_y > feature_map_height)
+                                    {
+                                        // For this sample we save 4x point (0,0) with weight 0
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
+                                        pooling_points.push_back({0, 0});
```
Comment (tomdol) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:106]: I recommend using the insert method instead https://en.cppreference.com/w/cpp/container/vector/insert 

`pooling_points.insert(pooling_points.end(), 4, {0, 0})`

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 2]
```diff
@@ -0,0 +1,243 @@
...
+
+                                        pooling_weights.push_back(0);
+                                        pooling_weights.push_back(0);
+                                        pooling_weights.push_back(0);
+                                        pooling_weights.push_back(0);
+
+                                        continue;
+                                    }
+
+                                    sample_x = std::max(sample_x, static_cast<T>(0.0));
```
Comment (tomdol) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:116]: You can just use `T{0}` instead of the `static_cast` here.

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+                        for (int64_t y_bin_ind = 0; y_bin_ind < pooled_height; y_bin_ind++)
+                        {
+                            for (int64_t x_bin_ind = 0; x_bin_ind < pooled_width; x_bin_ind++)
+                            {
+                                T pooled_value = 0;
+                                for (int64_t bin_sample_ind = 0;
+                                     bin_sample_ind < num_samples_in_bin;
+                                     bin_sample_ind++)
+                                {
+                                    auto s1 = feature_maps[feature_maps_transform.index(
```
Comment (tomdol) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:178]: I think the s1..s4 variables should be changed to something more elaborate.

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+                                    }
+                                    sample_index += 4;
+                                }
+                                tmp_out.push_back(pooled_value);
+                            }
+                        }
+                        // save the calculations for all bins across this channel
+                        std::copy(tmp_out.begin(),
+                                  tmp_out.end(),
+                                  &out[out_transform.index({static_cast<uint64_t>(roi_index),
```
Comment (tomdol) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:231]: Can you move the index method call somewhere above and assign it to a variable called `offset`? Then you can use this variable in the call to `std::copy` like `out + offset`.

File: ngraph/test/op_eval/roi_align.cpp [hunk 5]
```diff
@@ -0,0 +1,228 @@
...
+        33.4375f, 33.4688f, 35.1042f, 35.1042f, 35.1042f, 35.1354f, 56.7708f, 56.7708f, 56.7708f,
+        56.8021f, 58.4375f, 58.4375f, 58.4375f, 58.4688f, 60.1042f, 60.1042f, 60.1042f, 60.1354f};
+    const auto expected_shape = Shape{num_rois, C, pooled_height, pooled_width};
+
+    EXPECT_EQ(result->get_element_type(), element::f32);
+    EXPECT_EQ(result->get_shape(), expected_shape);
+
+    auto result_data = read_vector<float>(result);
+    for (size_t i = 0; i < expected_vec.size(); i++)
+        EXPECT_NEAR(result_data[i], expected_vec[i], 0.001);
```
Comment (tomdol) [inline ngraph/test/op_eval/roi_align.cpp:100]: We have an utility function that we use to compare the values in the tests called `all_close_f`. You should use that instead. Here's an example: https://github.com/openvinotoolkit/openvino/blob/master/ngraph/test/op_eval/matmul.cpp#L105

**Review by tomdol — CHANGES_REQUESTED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 1 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp]: I did change it here, left the push_backs that occur later, because i`m worried about decreasing readability when the inserted values are being calculated

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 2 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp]: Done

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 3 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp]: Changed the variable names, also added a comment above to maybe clarify what those variables are used for.

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 4 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp]: Done

File: ngraph/test/op_eval/roi_align.cpp (hunk 5 above)
Comment (blesniewski) [reply ngraph/test/op_eval/roi_align.cpp]: Done, but struggled a bit with choosing the correct parameters (In my opinion setting the max difference in EXPECT_NEAR is more user friendly than setting the bits in all_close_f)

**Review by tomdol — APPROVED** [review]

**Comment by blesniewski** [comment]
@postrational would you take a look at this pr?

**Review by jdanieck — COMMENTED** [review]
I would also add at least one test with simple model containing ROIAlign operator and execute it on INTERPRETER backend.  This will prove that it is working from user perspective.

Maybe we could use model from example in [OpenVINO](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/detection/ROIAlign_3.md) documentation.

**Review by jdanieck — APPROVED** [review]

**Review by postrational — APPROVED** [review]

**Review by sdurawa — APPROVED** [review]

File: ngraph/core/src/op/roi_align.cpp [hunk 6]
```diff
@@ -190,4 +202,104 @@ namespace ngraph
...
+        auto feature_maps = args[0];
+        auto rois = args[1];
+        auto batch_indices = args[2];
+
+        std::vector<int64_t> batch_indices_vec_scaled_up =
+            host_tensor_2_vector<int64_t>(batch_indices);
+
+        switch (feature_maps->get_element_type())
+        {
+        case element::Type_t::bf16:
```
Comment (mbencer) [inline ngraph/core/src/op/roi_align.cpp:225]: Optinal: You can use `TYPE_CASE` macro in order to reduce number of lines. Look for example to https://github.com/openvinotoolkit/openvino/blob/cd722d72dfe334bc426e06e5cdf81f43841ecaf3/ngraph/core/src/op/multiply.cpp#L71

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+                    T x2 = rois[rois_transform.index({roi_index, 2})] * spatial_scale;
+                    T y2 = rois[rois_transform.index({roi_index, 3})] * spatial_scale;
+
+                    T roi_width = std::max(x2 - x1, static_cast<T>(1.0));
+                    T roi_height = std::max(y2 - y1, static_cast<T>(1.0));
+
+                    T bin_width = roi_width / pooled_width;
+                    T bin_height = roi_height / pooled_height;
+
+                    uint64_t num_samples_in_bin = sampling_ratio * sampling_ratio;
```
Comment (mbencer) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:68]: In spec there is described case when `sampling_ratio == 0`. In such case we should use formula `ceil(roi_height / pooled_h)` and `ceil(roi_width / pooled_w)`

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+
+                                    auto sample_y_low = static_cast<unsigned int>(sample_y);
+                                    auto sample_x_low = static_cast<unsigned int>(sample_x);
+                                    unsigned int sample_y_high;
+                                    unsigned int sample_x_high;
+
+                                    if (sample_y_low >= feature_map_height - 1)
+                                    {
+                                        sample_y_high = sample_y_low = feature_map_height - 1;
+                                        sample_y = (T)sample_y_low;
```
Comment (mbencer) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:119]: Optional: earlier is used `static_case` style, I think that is better to keep uniformity

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+                                 y_sample_ind++)
+                            {
+                                T sample_y = y1 + static_cast<T>(y_bin_ind) * bin_height +
+                                             sample_distance_y * (static_cast<T>(y_sample_ind) +
+                                                                  static_cast<T>(0.5f));
+
+                                for (int64_t x_sample_ind = 0; x_sample_ind < sampling_ratio;
+                                     x_sample_ind++)
+                                {
+                                    T sample_x = x1 + static_cast<T>(x_bin_ind) * bin_width +
```
Comment (mbencer) [inline ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:95]: Can you please indicate source of used formulas? In spec there is only "bilinear interpolation" keyword

**Review by mbencer — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/roi_align.cpp (hunk 6 above)
Comment (blesniewski) [reply ngraph/core/src/op/roi_align.cpp:225]: Yeah, I really don't like this macro

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 7 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:68]: You`re right

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 8 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:119]: I agree here, somehow this didn't bother me before

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 9 above)
Comment (blesniewski) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:95]: So the formulas for sample_x and sample_y are rather easy to figure out- we know that in each bin we need to take sampling_ratio * sampling_ratio samples, but how to figure out where they should be located? 
Well a few articles about ROI align described this in a similar fashion, but some differences:
-  for horizontal spacing just divide the bin width by sampling ratio, then space the first sample out by half the value of horizontal spacing - no link for this article as I don't have any on hand
-  for horizontal spacing just divide the bin width by sampling ratio+1, then space the first sample out by the value of horizontal spacing: [article](https://towardsdatascience.com/understanding-region-of-interest-part-2-roi-align-and-roi-warp-f795196fc193)

The rest of line 95 is just starting from ROI`s beginning and then shifting to appropriate bin.

**Comment by ilyachur** [comment]
Could you add measurements about impact on binary size?

**Comment by blesniewski** [comment]
@ilyachur 
Sure thing, so on a release version:
Before: 7714952B (7.3576MB) libngraph.so
After:   7747720B (7.3888MB) libngraph.so
Impact on binary size: +32 768B (32KB)

File: ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp (hunk 9 above)
Comment (mbencer) [reply ngraph/core/reference/include/ngraph/runtime/reference/roi_align.hpp:100]: thanks, especially [article](https://towardsdatascience.com/understanding-region-of-interest-part-2-roi-align-and-roi-warp-f795196fc193) is very good ;)

**Review by mbencer — APPROVED** [review]

File: ngraph/core/include/ngraph/op/roi_align.hpp [hunk 10]
```diff
@@ -76,6 +76,9 @@ namespace ngraph
                 int get_sampling_ratio() const { return m_sampling_ratio; }
                 float get_spatial_scale() const { return m_spatial_scale; }
                 PoolingMode get_mode() const { return m_mode; }
+                bool evaluate(const HostTensorVector& outputs,
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/roi_align.hpp:79]: Just the general question. At the current moment we have a recommendation to don't add evaluate method to operation if we use it only in tests. Do we have real topology where we need to call evaluate for this topology?

File: ngraph/core/src/op/roi_align.cpp [hunk 11]
```diff
@@ -190,4 +202,104 @@ namespace ngraph
...
+                                                 pooled_height,
+                                                 pooled_width,
+                                                 sampling_ratio,
+                                                 spatial_scale,
+                                                 pooling_mode);
+            break;
+        }
+        case element::Type_t::f64:
+        {
+            runtime::reference::roi_align<double>(feature_maps->get_data_ptr<double>(),
```
Comment (ilyachur) [inline ngraph/core/src/op/roi_align.cpp:278]: Do we need evaluate for `f64` data? IE plugins don't support such precision and we don't add it to evaluate methods in order to decrease the impact on binary size.

File: ngraph/frontend/onnx_import/src/op/roi_align.cpp [hunk 12]
```diff
@@ -39,8 +39,8 @@ namespace ngraph
                     const auto& rois = inputs[1];
                     const auto& num_rois = inputs[2];
 
-                    const auto pooled_h = node.get_attribute_value<int64_t>("output_height", 1);
-                    const auto pooled_w = node.get_attribute_value<int64_t>("output_width", 1);
+                    const auto pooled_h = node.get_attribute_value<int64_t>("pooled_h", 1);
+                    const auto pooled_w = node.get_attribute_value<int64_t>("pooled_w", 1);
```
Comment (ilyachur) [inline ngraph/frontend/onnx_import/src/op/roi_align.cpp:43]: Why did you change the attribute names? Is it a bug which we have before?

Comment (blesniewski) [reply ngraph/frontend/onnx_import/src/op/roi_align.cpp:43]: Documentation specifies the attribute names to be pooled_h and pooled_w: [here](https://github.com/openvinotoolkit/openvino/blob/master/docs/ops/detection/ROIAlign_3.md)

File: ngraph/core/src/op/roi_align.cpp (hunk 11 above)
Comment (blesniewski) [reply ngraph/core/src/op/roi_align.cpp:278]: I will remove it then, documentation said the op should support all supported floating point types, so the fact that this type is present in for example `element::Type_t `suggested to me that it should be supported.

File: ngraph/core/include/ngraph/op/roi_align.hpp (hunk 10 above)
Comment (tomdol) [reply ngraph/core/include/ngraph/op/roi_align.hpp:79]: AFAIR this op's evaluator was requested by the VPU team for validation purposes.

File: ngraph/frontend/onnx_import/src/op/roi_align.cpp (hunk 12 above)
Comment (tomdol) [reply ngraph/frontend/onnx_import/src/op/roi_align.cpp:43]: This is a good catch - this code retrieves the attributes from an ONNX node which means you should follow this spec https://github.com/onnx/onnx/blob/master/docs/Operators.md#RoiAlign

File: ngraph/core/include/ngraph/op/roi_align.hpp (hunk 10 above)
Comment (blesniewski) [reply ngraph/core/include/ngraph/op/roi_align.hpp:79]: I think this op is required to support Mask-RCNN, there is a jira ticket for that: 23134.

File: ngraph/core/src/op/roi_align.cpp (hunk 11 above)
Comment (ilyachur) [reply ngraph/core/src/op/roi_align.cpp:278]: Ok thank you!

File: ngraph/frontend/onnx_import/src/op/roi_align.cpp (hunk 12 above)
Comment (ilyachur) [reply ngraph/frontend/onnx_import/src/op/roi_align.cpp:43]: Ok, looks like we should revert this change.

**Review by ggalieroc — APPROVED** [review]

File: ngraph/frontend/onnx_import/src/op/roi_align.cpp (hunk 12 above)
Comment (blesniewski) [reply ngraph/frontend/onnx_import/src/op/roi_align.cpp]: I`ve reverted it

**Review by ilyachur — APPROVED** [review]

**Comment by blesniewski** [comment]
@openvinotoolkit/openvino-ngraph-maintainers this PR can be merged
